In [14]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

telco = pd.read_excel('Telco_customer_churn.xlsx')
display(telco)


,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,2569-WGERO,1,United States,California,Landers,92285,"34.341737, -116.539416",34.341737,-116.539416,Female,...,Two year,Yes,Bank transfer (automatic),21.15,1419.4,No,0,45,5306,NaN
7039,6840-RESVB,1,United States,California,Adelanto,92301,"34.667815, -117.536183",34.667815,-117.536183,Male,...,One year,Yes,Mailed check,84.80,1990.5,No,0,59,2140,NaN
7040,2234-XADUH,1,United States,California,Amboy,92304,"34.559882, -115.637164",34.559882,-115.637164,Female,...,One year,Yes,Credit card (automatic),103.20,7362.9,No,0,71,5560,NaN
7041,4801-JZAZL,1,United States,California,Angelus Oaks,92305,"34.1678, -116.86433",34.167800,-116.864330,Female,...,Month-to-month,Yes,Electronic check,29.60,346.45,No,0,59,2793,NaN


In [16]:
telco.columns

Index(['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code',
       'Lat Long', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen',
       'Partner', 'Dependents', 'Tenure Months', 'Phone Service',
       'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method',
       'Monthly Charges', 'Total Charges', 'Churn Label', 'Churn Value',
       'Churn Score', 'CLTV', 'Churn Reason'],
      dtype='str')

### Renomear colunas (opcional)

In [17]:
# Criando dicionário de tradução
traducao = {
    'CustomerID': 'ID_Cliente',
    'Count': 'Contagem',
    'Country': 'Pais',
    'State': 'Estado',
    'City': 'Cidade',
    'Zip Code': 'CEP',
    'Lat Long': 'Lat_Long',
    'Latitude': 'Latitude',
    'Longitude': 'Longitude',
    'Gender': 'Genero',
    'Senior Citizen': 'Idoso',
    'Partner': 'Parceiro',
    'Dependents': 'Dependentes',
    'Tenure Months': 'Meses_Contrato',
    'Phone Service': 'Servico_Telefonico',
    'Multiple Lines': 'Multiplas_Linhas',
    'Internet Service': 'Servico_Internet',
    'Online Security': 'Seguranca_Online',
    'Online Backup': 'Backup_Online',
    'Device Protection': 'Protecao_Dispositivo',
    'Tech Support': 'Suporte_Tecnico',
    'Streaming TV': 'Streaming_TV',
    'Streaming Movies': 'Streaming_Filmes',
    'Contract': 'Contrato',
    'Paperless Billing': 'Fatura_Digital',
    'Payment Method': 'Metodo_Pagamento',
    'Monthly Charges': 'Valor_Mensal',
    'Total Charges': 'Valor_Total',
    'Churn Label': 'Churn_Label',
    'Churn Value': 'Churn_Valor',
    'Churn Score': 'Pontuacao_Churn',
    'CLTV': 'CLTV',
    'Churn Reason': 'Motivo_Churn'
}

# Aplicando a renomeação
telco = telco.rename(columns=traducao)


### Verificar possíveis tratamentos de valores das colunas

In [18]:
#telco.info() #verificado que a coluna Valor_Total precisa ser transformada de str para float64
#print(telco.isnull().sum()) #foi verificado que não há linhas vazias em nenhuma das colunas

# Converter a coluna Valor_Total em numeric 
telco['Valor_Total'] = pd.to_numeric(telco['Valor_Total'], errors='coerce')

# Verificar o tipo
print(telco['Valor_Total'].dtypes)

float64


### Criando um dataframe para fazer Análise de Dados

In [20]:
# Colunas para fazer análises
colunas_analise = [
    'Cidade', 'CEP', 'Genero', 'Idoso', 'Parceiro', 'Dependentes', 
    'Meses_Contrato', 'Servico_Telefonico', 'Multiplas_Linhas', 
    'Servico_Internet', 'Seguranca_Online', 'Backup_Online', 
    'Protecao_Dispositivo', 'Suporte_Tecnico', 'Streaming_TV', 
    'Streaming_Filmes', 'Contrato', 'Fatura_Digital', 
    'Metodo_Pagamento', 'Valor_Mensal', 'Valor_Total', 
    'Churn_Label', 'CLTV', 'Motivo_Churn'
]

# Criando o novo DF
telco_analise = telco[colunas_analise].copy()
telco_analise.head()

,Cidade,CEP,Genero,Idoso,Parceiro,Dependentes,Meses_Contrato,Servico_Telefonico,Multiplas_Linhas,Servico_Internet,...,Streaming_TV,Streaming_Filmes,Contrato,Fatura_Digital,Metodo_Pagamento,Valor_Mensal,Valor_Total,Churn_Label,CLTV,Motivo_Churn
0,Los Angeles,90003,Male,No,No,No,2,Yes,No,DSL,...,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,3239,Competitor made better offer
1,Los Angeles,90005,Female,No,No,Yes,2,Yes,No,Fiber optic,...,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,2701,Moved
2,Los Angeles,90006,Female,No,No,Yes,8,Yes,Yes,Fiber optic,...,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,Yes,5372,Moved
3,Los Angeles,90010,Female,No,Yes,Yes,28,Yes,Yes,Fiber optic,...,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,5003,Moved
4,Los Angeles,90015,Male,No,No,Yes,49,Yes,Yes,Fiber optic,...,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,Yes,5340,Competitor had better devices


### Análises

- Total de venda por cidade

In [30]:
telco_vendas_cidade = telco_analise.groupby('Cidade')['Valor_Total'].sum().sort_values(ascending=False)
print(telco_vendas_cidade)

Cidade
Los Angeles      647751.25
San Diego        354896.60
Sacramento       256295.05
San Jose         243735.55
San Francisco    221624.65
                   ...    
Homeland            668.25
Dana Point          556.80
Loleta              484.65
Truckee             479.35
Eagleville          203.40
Name: Valor_Total, Length: 1129, dtype: float64


- Total vendas por metodo de pagamento

In [29]:
telco_vendas_metodo = telco_analise.groupby('Metodo_Pagamento')['Valor_Total'].sum().sort_values(ascending=False)
print(telco_vendas_metodo)

Metodo_Pagamento
Electronic check             4944903.25
Bank transfer (automatic)    4748279.90
Credit card (automatic)      4671593.35
Mailed check                 1691392.20
Name: Valor_Total, dtype: float64


- Total vendas por tipo de contrato

In [32]:
telco_vendas_contrato = telco_analise.groupby('Contrato')['Valor_Total'].sum().sort_values(ascending=False)
print(telco_vendas_contrato)

Contrato
Two year          6283253.7
Month-to-month    5305861.5
One year          4467053.5
Name: Valor_Total, dtype: float64


- Total vendas por tipo de serviço

In [34]:
telco_vendas_serviço = telco_analise.groupby('Servico_Internet')['Valor_Total'].sum().sort_values(ascending=False)
print(telco_vendas_serviço)

Servico_Internet
Fiber optic    9923622.95
DSL            5121410.85
No             1011134.90
Name: Valor_Total, dtype: float64


- Total vendas por genero

In [36]:
telco_vendas_genero = telco_analise.groupby('Genero')['Valor_Total'].sum().sort_values(ascending=False)
print(telco_vendas_genero)

Genero
Male      8103814.5
Female    7952354.2
Name: Valor_Total, dtype: float64


- Contagem de churn e motivo de churn

In [ ]:
telco_churn = telco_analise['Churn_Label'].value_counts()
print(telco_churn)
print('-'*20)
telco_motivo_churn = telco_analise['Motivo_Churn'].value_counts()
print(telco_motivo_churn)

Churn_Label
No     5174
Yes    1869
Name: count, dtype: int64
--------------------
Motivo_Churn
Attitude of support person                   192
Competitor offered higher download speeds    189
Competitor offered more data                 162
Don't know                                   154
Competitor made better offer                 140
Attitude of service provider                 135
Competitor had better devices                130
Network reliability                          103
Product dissatisfaction                      102
Price too high                                98
Service dissatisfaction                       89
Lack of self-service on Website               88
Extra data charges                            57
Moved                                         53
Limited range of services                     44
Lack of affordable download/upload speed      44
Long distance charges                         44
Poor expertise of phone support               20
Poor expertise of onli

- Atitude do atendente de suporte              192
- Concorrente ofereceu maior velocidade de download    189
- Concorrente ofereceu mais dados             162
- Não sabe                                    154
- Concorrente fez uma oferta melhor           140
- Atitude do prestador de serviço             135
- Concorrente tinha melhores dispositivos     130
- Confiabilidade da rede                      103
- Insatisfação com o produto                  102
- Preço muito alto                             98
- Insatisfação com o serviço                   89
- Falta de autoatendimento no site             88
- Cobranças extras por dados                   57
- Mudou-se                                     53
- Gama limitada de serviços                    44
- Falta de velocidade de download/upload acessível   44
- Cobranças de longa distância                 44
- Baixa expertise do suporte telefônico        20
- Baixa expertise do suporte online            19
- Falecido                                      6


- Churn é com CLTV alto ou baixo?

In [42]:
telco_churn_alto_baixo = telco_analise.groupby('Churn_Label')['CLTV'].mean()
print(telco_churn_alto_baixo)


Churn_Label
No     4490.921337
Yes    4149.414660
Name: CLTV, dtype: float64


- Churn por tipo de contrato

In [44]:
pd.crosstab(telco_analise['Contrato'], telco_analise['Churn_Label'], normalize='index')

Churn_Label,No,Yes
Contrato,,
Month-to-month,0.572903,0.427097
One year,0.887305,0.112695
Two year,0.971681,0.028319


Quanto mais longo o contrato, menor o churn

- Churn e metodo de pagamento

In [48]:
pd.crosstab(telco_analise['Metodo_Pagamento'], telco_analise['Churn_Label'], normalize='index')


Churn_Label,No,Yes
Metodo_Pagamento,,
Bank transfer (automatic),0.832902,0.167098
Credit card (automatic),0.847569,0.152431
Electronic check,0.547146,0.452854
Mailed check,0.808933,0.191067


Clientes que usam electronic check têm uma probabilidade muito maior de sair.

- Cruzar método de pagamento com contrato

In [56]:
pd.crosstab(telco_analise['Metodo_Pagamento'], telco_analise['Contrato'], normalize='index')

Contrato,Month-to-month,One year,Two year
Metodo_Pagamento,,,
Bank transfer (automatic),0.381477,0.253238,0.365285
Credit card (automatic),0.356767,0.261498,0.381735
Electronic check,0.782241,0.146723,0.071036
Mailed check,0.553970,0.209057,0.236973


Na análise anterior, observou-se que clientes que utilizam o método de pagamento electronic check apresentam maior probabilidade de churn, assim como clientes com contrato do tipo month-to-month.
Esta análise confirma essa relação, mostrando que aproximadamente 78% dos clientes que utilizam electronic check estão em contratos mensais, o que indica um perfil de maior risco de cancelamento.

- Motivo de churn e impacto financeiro

In [54]:
motivo_churn_tabela = (
    telco_analise[telco_analise['Churn_Label'] == 'Yes']
    .groupby('Motivo_Churn')['Valor_Total']
    .sum()
    .sort_values(ascending=False)
    .reset_index()
)

display(motivo_churn_tabela)


,Motivo_Churn,Valor_Total
0,Competitor offered higher download speeds,318334.95
1,Attitude of support person,294783.20
2,Competitor offered more data,273004.90
3,Attitude of service provider,214957.80
4,Don't know,208784.20
5,Competitor made better offer,198153.55
6,Product dissatisfaction,185742.25
7,Competitor had better devices,182545.80
8,Price too high,149443.70
9,Network reliability,142756.30


- Competidor ofereceu maior velocidade de download    318334.95
- Atitude do atendente de suporte                    294783.20
- Competidor ofereceu mais dados                    273004.90
- Atitude do prestador de serviço                   214957.80
- Não sabe                                          208784.20
- Competidor fez uma oferta melhor                  198153.55
- Insatisfação com o produto                        185742.25
- Competidor tinha melhores dispositivos            182545.80
- Preço muito alto                                  149443.70
- Confiabilidade da rede                           142756.30
- Insatisfação com o serviço                       137534.25
- Cobranças extras de dados                        113798.85
- Falta de autoatendimento no site                106965.75
- Mudou-se                                          87020.15
- Falta de velocidade acessível de download/upload  74937.35
- Cobranças de longa distância                     74250.30
- Gama limitada de serviços                        43275.70
- Baixa competência do suporte telefônico         30513.95
- Baixa competência do suporte online             22829.85
- Falecido                                          3294.10


- Churn por cidade
1. cidades com base mínima

In [61]:
cidades_validas = telco_analise['Cidade'].value_counts()
cidades_validas = cidades_validas[cidades_validas >= 30].index

telco_filtrado = telco_analise[telco_analise['Cidade'].isin(cidades_validas)]
print(telco_filtrado)

           Cidade    CEP  Genero Idoso Parceiro Dependentes  Meses_Contrato  \
0     Los Angeles  90003    Male    No       No          No               2   
1     Los Angeles  90005  Female    No       No         Yes               2   
2     Los Angeles  90006  Female    No       No         Yes               8   
3     Los Angeles  90010  Female    No      Yes         Yes              28   
4     Los Angeles  90015    Male    No       No         Yes              49   
...           ...    ...     ...   ...      ...         ...             ...   
7001    San Diego  92128    Male    No       No         Yes              31   
7002    San Diego  92129    Male    No      Yes         Yes              37   
7003    San Diego  92131  Female   Yes       No          No              35   
7004    San Diego  92139  Female    No      Yes         Yes              23   
7005    San Diego  92154  Female    No      Yes         Yes              20   

     Servico_Telefonico  Multiplas_Linhas Servico_I

- Churn real

In [62]:
pd.crosstab(
    telco_filtrado['Cidade'],
    telco_filtrado['Churn_Label'],
    normalize='index'
)

Churn_Label,No,Yes
Cidade,,
Bakersfield,0.925000,0.075000
Berkeley,0.718750,0.281250
Fresno,0.750000,0.250000
Glendale,0.675000,0.325000
Long Beach,0.750000,0.250000
Los Angeles,0.704918,0.295082
Oakland,0.750000,0.250000
Pasadena,0.700000,0.300000
Riverside,0.750000,0.250000


In [66]:
#telco_analise.isnull().sum()
#Ajustando valores nulos no Motivo_Churn
telco_analise['Motivo_Churn'] = telco_analise['Motivo_Churn'].fillna('Não churn') 
#verificando linhas com valor faltante
telco_analise[telco_analise['Valor_Total'].isnull()]
# Preenchendo valores faltantes (meses contrato * valor_mensal)
telco_analise['Valor_Total'] = telco_analise['Valor_Total'].fillna(
    telco_analise['Valor_Mensal'] * telco_analise['Meses_Contrato']
)

- Nummero de serviços contratados

In [69]:
servicos = [
    'Servico_Telefonico',
    'Multiplas_Linhas',
    'Servico_Internet',
    'Seguranca_Online',
    'Backup_Online',
    'Protecao_Dispositivo',
    'Suporte_Tecnico',
    'Streaming_TV',
    'Streaming_Filmes'
]

telco_analise['Qtd_Servicos'] = telco_analise[servicos].apply(lambda x: sum(x == 'Yes'), axis=1)

- Cliente com risco de churn

In [71]:
telco_analise['Alto_Risco'] = (
    (telco_analise['Contrato'] == 'Month-to-month') &
    (telco_analise['Metodo_Pagamento'] == 'Electronic check')
)

- Quantidade de clientes com risco de churn

In [80]:
print(telco_analise['Alto_Risco'].sum())
print(telco_analise['Alto_Risco'].mean() * 100)



1850
26.267215675138434


In [81]:
telco_analise.head()
telco_analise.columns


Index(['Cidade', 'CEP', 'Genero', 'Idoso', 'Parceiro', 'Dependentes',
       'Meses_Contrato', 'Servico_Telefonico', 'Multiplas_Linhas',
       'Servico_Internet', 'Seguranca_Online', 'Backup_Online',
       'Protecao_Dispositivo', 'Suporte_Tecnico', 'Streaming_TV',
       'Streaming_Filmes', 'Contrato', 'Fatura_Digital', 'Metodo_Pagamento',
       'Valor_Mensal', 'Valor_Total', 'Churn_Label', 'CLTV', 'Motivo_Churn',
       'Qtd_Servicos', 'Alto_Risco'],
      dtype='str')

### Preparação para testar modelos

In [82]:
df = telco_analise.copy()

# Target
y = df['Churn_Label'].map({'Yes': 1, 'No': 0})

# Remover colunas que vazam informação (leakage)
drop_cols = ['Churn_Label', 'Motivo_Churn']
X = df.drop(columns=drop_cols)

#separar colunas numéricas e texto
cat_cols = X.select_dtypes(include='object').columns
num_cols = X.select_dtypes(exclude='object').columns

C:\Users\fwfco\AppData\Local\Temp\ipykernel_20720\3488502967.py:11: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = X.select_dtypes(include='object').columns


- Modelos

In [83]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

preprocess = ColumnTransformer(
    transformers=[
        ('num', SimpleImputer(strategy='constant', fill_value=0), num_cols),
        ('cat', Pipeline(steps=[
            ('imputer', SimpleImputer(strategy='most_frequent')),
            ('onehot', OneHotEncoder(handle_unknown='ignore'))
        ]), cat_cols)
    ]
)



### Treino

In [84]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

### Modelos
### Decision Tree

In [92]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report

tree_model = Pipeline(steps=[
    ('preprocess', preprocess),
    ('model', DecisionTreeClassifier(
        random_state=42,
        class_weight='balanced'
    ))
])

tree_model.fit(X_train, y_train)
pred_tree = tree_model.predict(X_test)

print("Decision Tree")
print(classification_report(y_test, pred_tree))

Decision Tree
              precision    recall  f1-score   support

           0       0.85      0.81      0.83      1035
           1       0.53      0.60      0.56       374

    accuracy                           0.75      1409
   macro avg       0.69      0.70      0.69      1409
weighted avg       0.76      0.75      0.76      1409



### Random Forest

In [93]:
from sklearn.ensemble import RandomForestClassifier

rf_model = Pipeline(steps=[
    ('preprocess', preprocess),
    ('model', RandomForestClassifier(
        n_estimators=300,
        random_state=42,
        class_weight='balanced'
    ))
])

rf_model.fit(X_train, y_train)
pred_rf = rf_model.predict(X_test)

print("Random Forest")
print(classification_report(y_test, pred_rf))

Random Forest
              precision    recall  f1-score   support

           0       0.84      0.89      0.86      1035
           1       0.63      0.51      0.57       374

    accuracy                           0.79      1409
   macro avg       0.73      0.70      0.71      1409
weighted avg       0.78      0.79      0.78      1409



### XGBoost

In [94]:
from xgboost import XGBClassifier

xgb_model = Pipeline(steps=[
    ('preprocess', preprocess),
    ('model', XGBClassifier(
        n_estimators=300,
        learning_rate=0.05,
        max_depth=5,
        scale_pos_weight=(len(y_train[y_train==0]) / len(y_train[y_train==1])),
        eval_metric='logloss',
        random_state=42
    ))
])

xgb_model.fit(X_train, y_train)
pred_xgb = xgb_model.predict(X_test)

print("XGBoost")
print(classification_report(y_test, pred_xgb))

XGBoost
              precision    recall  f1-score   support

           0       0.90      0.76      0.82      1035
           1       0.53      0.77      0.63       374

    accuracy                           0.76      1409
   macro avg       0.72      0.76      0.73      1409
weighted avg       0.80      0.76      0.77      1409



### Comparação

In [97]:
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score

def evaluate(name, y_true, y_pred):
    print(f"\n{name}")
    print("Accuracy:", accuracy_score(y_true, y_pred))
    print("Recall (Churn):", recall_score(y_true, y_pred))
    print("Precision (Churn):", precision_score(y_true, y_pred))
    print("F1 (Churn):", f1_score(y_true, y_pred))

evaluate("Decision Tree", y_test, pred_tree)
evaluate("Random Forest", y_test, pred_rf)
evaluate("XGBoost", y_test, pred_xgb)


Decision Tree
Accuracy: 0.751596877217885
Recall (Churn): 0.5989304812834224
Precision (Churn): 0.5283018867924528
F1 (Churn): 0.5614035087719298

Random Forest
Accuracy: 0.7906316536550745
Recall (Churn): 0.5133689839572193
Precision (Churn): 0.6295081967213115
F1 (Churn): 0.5655375552282769

XGBoost
Accuracy: 0.759403832505323
Recall (Churn): 0.7700534759358288
Precision (Churn): 0.532347504621072
F1 (Churn): 0.6295081967213115


In [96]:
from sklearn.metrics import classification_report

print(classification_report(y_test, pred_rf))

              precision    recall  f1-score   support

           0       0.84      0.89      0.86      1035
           1       0.63      0.51      0.57       374

    accuracy                           0.79      1409
   macro avg       0.73      0.70      0.71      1409
weighted avg       0.78      0.79      0.78      1409



In [98]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier

from xgboost import XGBClassifier

from sklearn.metrics import classification_report

# -----------------------------
# 1. Base
# -----------------------------
df = telco_analise.copy()

y = df['Churn_Label'].map({'Yes': 1, 'No': 0})

# -----------------------------
# 2. Feature Engineering
# -----------------------------
df['Ticket_Medio'] = df['Valor_Total'] / df['Meses_Contrato']

df['Qtd_Servicos'] = df[
    ['Servico_Telefonico','Multiplas_Linhas','Servico_Internet',
     'Seguranca_Online','Backup_Online','Protecao_Dispositivo',
     'Suporte_Tecnico','Streaming_TV','Streaming_Filmes']
].apply(lambda x: sum(x == 'Yes'), axis=1)

df['Alto_Risco'] = (
    (df['Contrato'] == 'Month-to-month') &
    (df['Metodo_Pagamento'] == 'Electronic check')
)

# -----------------------------
# 3. Features
# -----------------------------
X = df.drop(columns=['Churn_Label', 'Motivo_Churn'])

cat_cols = X.select_dtypes(include=['object']).columns
num_cols = X.select_dtypes(exclude=['object']).columns

# -----------------------------
# 4. Preprocessamento
# -----------------------------
preprocess = ColumnTransformer(
    transformers=[
        ('num', Pipeline(steps=[
            ('imputer', SimpleImputer(strategy='constant', fill_value=0)),
            ('scaler', StandardScaler())
        ]), num_cols),

        ('cat', Pipeline(steps=[
            ('imputer', SimpleImputer(strategy='most_frequent')),
            ('onehot', OneHotEncoder(handle_unknown='ignore'))
        ]), cat_cols)
    ]
)

# -----------------------------
# 5. Split
# -----------------------------
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# -----------------------------
# 6. Modelos
# -----------------------------
scale_pos_weight = (y_train == 0).sum() / (y_train == 1).sum()

models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),

    "KNN": KNeighborsClassifier(n_neighbors=5),

    "Decision Tree": DecisionTreeClassifier(
        class_weight='balanced',
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=300,
        class_weight='balanced',
        random_state=42
    ),

    "Gradient Boosting": GradientBoostingClassifier(),

    "XGBoost": XGBClassifier(
        n_estimators=400,
        learning_rate=0.05,
        max_depth=5,
        subsample=0.8,
        colsample_bytree=0.8,
        scale_pos_weight=scale_pos_weight,
        eval_metric='logloss',
        random_state=42
    )
}

# -----------------------------
# 7. Treino + Avaliação
# -----------------------------
for name, clf in models.items():
    model = Pipeline(steps=[
        ('preprocess', preprocess),
        ('model', clf)
    ])
    
    model.fit(X_train, y_train)
    preds = model.predict(X_test)
    
    print("\n====================")
    print(name)
    print("====================")
    print(classification_report(y_test, preds))

C:\Users\fwfco\AppData\Local\Temp\ipykernel_20720\1569130504.py:46: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = X.select_dtypes(include=['object']).columns



Logistic Regression
              precision    recall  f1-score   support

           0       0.84      0.87      0.86      1035
           1       0.61      0.55      0.58       374

    accuracy                           0.79      1409
   macro avg       0.73      0.71      0.72      1409
weighted avg       0.78      0.79      0.78      1409


KNN
              precision    recall  f1-score   support

           0       0.84      0.85      0.84      1035
           1       0.57      0.56      0.56       374

    accuracy                           0.77      1409
   macro avg       0.70      0.70      0.70      1409
weighted avg       0.77      0.77      0.77      1409


Decision Tree
              precision    recall  f1-score   support

           0       0.83      0.82      0.83      1035
           1       0.52      0.54      0.53       374

    accuracy                           0.75      1409
   macro avg       0.68      0.68      0.68      1409
weighted avg       0.75      0.75